sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala

```diff
@@ -3422,6 +3422,74 @@ object GpuOverrides extends Logging {
+            TypeSig.DOUBLE + TypeSig.ARRAY.nested(TypeSig.DOUBLE)),
+          ParamCheck("frequency",
+            TypeSig.LONG + TypeSig.ARRAY.nested(TypeSig.LONG),
+            TypeSig.LONG + TypeSig.ARRAY.nested(TypeSig.LONG)))),
+      (c, conf, p, r) => new TypedImperativeAggExprMeta[Percentile](c, conf, p, r) {
+        override def tagAggForGpu(): Unit = {
+          // Check if the input percentage can be supported on GPU.
+          childExprs(1).wrapped match {
+            case lit: Literal => lit.value match {
+              case null =>
+                willNotWorkOnGpu(
+                  "percentile on GPU only supports non-null literal percentages")
+              case a: ArrayData if (0 until a.numElements).exists(a.isNullAt) =>
+                willNotWorkOnGpu(
+                  "percentile on GPU does not support percentage arrays containing nulls")
+              case a: ArrayData if a.toDoubleArray()
+                .exists(percentage => percentage < 0.0 || percentage > 1.0) =>
+                willNotWorkOnGpu(
+                  "percentile requires the input percentages given in the range [0, 1]")
+              case _ => // This is fine.
+            }
+            case _ =>
+              willNotWorkOnGpu("percentile on GPU only supports literal percentages")
+          }
+        }
+
+        override def convertToGpu(childExprs: Seq[Expression]): GpuExpression = {
+          val exprMeta = p.get.asInstanceOf[BaseExprMeta[_]]
+          val isReduction = exprMeta.context match {
+            case ReductionAggExprContext => true
+            case GroupByAggExprContext => false
+            case _ => throw new IllegalStateException(
+              s"Invalid aggregation context: ${exprMeta.context}")
+          }
+          GpuPercentile(childExprs.head, childExprs(1).asInstanceOf[GpuLiteral], childExprs(2),
+            isReduction)
+        }
+        // Declare the data type of the internal buffer so it can be serialized and
+        // deserialized correctly during shuffling.
+        override def aggBufferAttribute: AttributeReference = {
+          val aggBuffer = c.aggBufferAttributes.head
+          val dataType: DataType = ArrayType(StructType(Seq(
+            StructField("value", childExprs.head.dataType),
+            StructField("frequency", LongType))), containsNull = false)
+          aggBuffer.copy(dataType = dataType)(aggBuffer.exprId, aggBuffer.qualifier)
+        }
+
+        override val needsAnsiCheck: Boolean = false
+        override val supportBufferConversion: Boolean = true
+        override def createCpuToGpuBufferConverter(): CpuToGpuAggregateBufferConverter =
+          CpuToGpuPercentileBufferConverter(childExprs.head.dataType)
+        override def createGpuToCpuBufferConverter(): GpuToCpuAggregateBufferConverter =
+          GpuToCpuPercentileBufferConverter(childExprs.head.dataType)
+      }).incompat("the GPU implementation of percentile is not bit-for-bit " +
```

Comment:
Do we know how it is different? Do we want to file a follow on issue to deal with this? Is it just rounding errors in floating point?

Reply:
It is floating point round off error. And it is interesting that Spark CPU seems to have larger error than ours:
```
scala> df.selectExpr("percentile(value, 0.1)").show
+-----------------------------------------+
|percentile(value, CAST(0.1 AS DOUBLE), 1)|
+-----------------------------------------+
|                                    5.918|                <=== our
+-----------------------------------------+


scala> spark.conf.set("spark.rapids.sql.enabled", "false")

scala> df.selectExpr("percentile(value, 0.1)").show
+-----------------------------------------+
|percentile(value, CAST(0.1 AS DOUBLE), 1)|
+-----------------------------------------+
|                        5.918000000000001|
+-----------------------------------------+
```